implementations/rust/ockam/ockam_node/src/router/mod.rs

```diff
@@ -269,11 +276,13 @@ impl Router {
 
     /// Block current task running this router.  Return fatal errors
     async fn run_inner(&mut self) -> Result<()> {
-        while let Some(msg) = self.receiver.recv().await {
+        while let Some(msg) = self.get_recv()?.recv().await {
             let msg_str = format!("{}", msg);
             match self.handle_msg(msg).await {
                 Ok(should_break) => {
                     if should_break {
+                        // We drop the receiver end here
+                        self.receiver.take();
```

Comment:
Rather than dropping the receiver I recommend calling `Receiver::close` instead. This way there is no need to have it as an `Option`al field in `Router`.